src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.lang.stable;
+
+import jdk.internal.lang.StableValue;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+import static jdk.internal.lang.stable.StableUtil.*;
+
+public final class StableValueImpl<V> implements StableValue<V> {
+
+    private static final long MUTEX_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "mutex");
+
+    private static final long VALUE_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "value");
+
+    private static final long STATE_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "state");
+
+    /**
+     * An internal mutex used rather than synchronizing on `this`. Lazily created.
+     * If `null`    , we have not entered a mutex section yet
+     * If TOMBSTONE , we do not need synchronization anymore (isSet() && isError() = true)
+     * if instanceof Throwable , records a previous error and, we do not need synchronization anymore
+     * otherwise    , a distinct synchronization object
+     */
+    private Object mutex;
+
+    /**
+     * If non-null, holds a set value
+     * If `null`  , may be unset or hold a set `null` value
+     */
+    @Stable
+    private V value;
+
+    /**
+     * If StableUtil.NOT_SET  , a value is not set
+     * If StableUtil.NON_NULL , a non-null value is set
+     * If StableUtil.NULL     , a `null` value is set
+     */
+    @Stable
+    private int state;
```

Comment:
Can we change this to be a byte, so state and supplying fields can be packed together in 4 bytes in some layouts?

Reply:
We had `byte` before but converted to `int` as the `byte` will get promoted to an `int` anyhow in the code logic. However, one idea would be to go back to using a `byte` again and also use a `byte` for the flag of computation invocation. This would reduce the footprint for these two fields from 8 bytes to 2 bytes.